File: Leetcode/Wildcard Matching/Wildcard Matching.py
#Problem:https://leetcode.com/problems/wildcard-matching/
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        n = len(s) # row
        m = len(p) # column
        
        # dynamic programming table initialization
        
        dp = [ [False for i in range(m+1)] for j in range(n+1) ]
        
        # if both the text and the pattern is an empty string then it's a match
        dp[0][0] = True
        
        # this is for the 0th row where a kleene star can take 0 or any sequence of characters
        # hence we look at the cell to the left
        
        for i in range(1, m+1):
            if p[i-1] == '*':
                dp[0][i] = dp[0][i-1]
        
        for i in range(1, n+1):
            for j in range(1, m+1):
                # direct match hence we look at the cell diagonally above
                # if the substring is a match without the character then it is also
                # a match for the given string
                
                if s[i-1] == p[j-1] or p[j-1] == '?':
                    dp[i][j] = dp[i-1][j-1]

                # hence we look at the cell to the left and at top
                # if either of the cells return True then the string is a match
                # this is because we need to consider the 0th occurance -> cell to the left
                # and also any number of sequence -> cell at the top
                if p[j-1] == '*':
                    dp[i][j] = dp[i-1][j] or dp[i][j-1]
         
        return dp[-1][-1]  # this returns the last cell which has the answer

File: Leetcode/Wildcard Matching/Wildcard Matching.java
public boolean isMatch(String s, String p) {
    // replace multiple * with single *
    char[] sArray = s.toCharArray(), pArray = p.toCharArray();
    int writeIndex = 0;
    boolean isFirst = true;
    for(int i=0; i<pArray.length; i++) {
        if(pArray[i] == '*') {
            if(isFirst) {
                pArray[writeIndex++] = pArray[i];
                isFirst = false;
            }
        } else {
            pArray[writeIndex++] = pArray[i];
            isFirst = true;
        }
    }
    
    // initiate dp array
    int m = sArray.length, n = writeIndex;
    boolean[][] dp = new boolean[m+1][n+1];
    
    // s and p are all empty, they match
    dp[0][0] = true;
    
    // s is empty and p is *, they match
    if(n > 0 && pArray[0] == '*') dp[0][1] = true;
    
    for(int i=1; i<m+1; i++) {
        for(int j=0; j<n+1; j++) {
            // add j>0 to make first column false, because pattern is empty but text is not
            if(j > 0 && (sArray[i-1] == pArray[j-1] || pArray[j-1] == '?')) {
                dp[i][j] = dp[i-1][j-1];
            } else if(j > 0 && pArray[j-1] == '*') {
                dp[i][j] = dp[i-1][j] || dp[i][j-1];
            } else {
                dp[i][j] = false;
            }
        }
    }
    return dp[m][n];
}